### ISSUE TITLE
How can i get the cuda version of troch-scatter?

### ISSUE THREAD

**Issue by messyQ1128** [rusty1s/pytorch_geometric#623]
## ❓ Questions & Help

When i import the torch_geometric, there is an err occured 
which is " ModuleNotFoundError: No module named 'torch_scatter.scatter_cuda'"

How can i fix this?
Thanks alot!

My pytorch version is 1.2.0. (Installed with conda install pytorch torchvision cudatoolkit=10.0 -c pytorch)
My cuda version is 10.0.
My gcc version is gcc 7.4.0.

**Comment by rusty1s** [comment]
The installation of `torch-scatter` did not succeed. Please uninstall the package, follow the installation guide and post the installation log here.

**Comment by messyQ1128** [comment]
> The installation of `torch-scatter` did not succeed. Please uninstall the package, follow the installation guide and post the installation log here.


**when i have install torch-cluster, the err has occured.**
**i can't fix this,would you help me please?**

the err has been posted below

(base) wolf@wolf-Alienware-15-R4:~$ pip install torch-cluster
Collecting torch-cluster
  Using cached https://files.pythonhosted.org/packages/bd/5f/01c5799cd1f81f9956f03a0e1d9a861e020a598dd411d9bd3c3c1dd5b8a4/torch_cluster-1.4.4.tar.gz
Requirement already satisfied: scipy in ./anaconda3/lib/python3.7/site-packages (from torch-cluster) (1.3.0)
Building wheels for collected packages: torch-cluster
  Building wheel for torch-cluster (setup.py) ... error
  ERROR: Complete output from command /home/wolf/anaconda3/bin/python -u -c 'import setuptools, tokenize;__file__='"'"'/tmp/pip-install-19kdi0h_/torch-cluster/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-yrl7y1sx --python-tag cp37:
  ERROR: running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-x86_64-3.7
  creating build/lib.linux-x86_64-3.7/test
  copying test/__init__.py -> build/lib.linux-x86_64-3.7/test
  copying test/test_radius.py -> build/lib.linux-x86_64-3.7/test
  copying test/utils.py -> build/lib.linux-x86_64-3.7/test
  copying test/test_sampler.py -> build/lib.linux-x86_64-3.7/test
  copying test/test_rw.py -> build/lib.linux-x86_64-3.7/test
  copying test/test_fps.py -> build/lib.linux-x86_64-3.7/test
  copying test/test_knn.py -> build/lib.linux-x86_64-3.7/test
  copying test/test_grid.py -> build/lib.linux-x86_64-3.7/test
  copying test/test_nearest.py -> build/lib.linux-x86_64-3.7/test
  copying test/test_graclus.py -> build/lib.linux-x86_64-3.7/test
  creating build/lib.linux-x86_64-3.7/torch_cluster
  copying torch_cluster/__init__.py -> build/lib.linux-x86_64-3.7/torch_cluster
  copying torch_cluster/grid.py -> build/lib.linux-x86_64-3.7/torch_cluster
  copying torch_cluster/fps.py -> build/lib.linux-x86_64-3.7/torch_cluster
  copying torch_cluster/nearest.py -> build/lib.linux-x86_64-3.7/torch_cluster
  copying torch_cluster/knn.py -> build/lib.linux-x86_64-3.7/torch_cluster
  copying torch_cluster/rw.py -> build/lib.linux-x86_64-3.7/torch_cluster
  copying torch_cluster/sampler.py -> build/lib.linux-x86_64-3.7/torch_cluster
  copying torch_cluster/graclus.py -> build/lib.linux-x86_64-3.7/torch_cluster
  copying torch_cluster/radius.py -> build/lib.linux-x86_64-3.7/torch_cluster
  running build_ext
  building 'torch_cluster.graclus_cpu' extension
  creating build/temp.linux-x86_64-3.7
  creating build/temp.linux-x86_64-3.7/cpu
  gcc -pthread -B /home/wolf/anaconda3/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include -I/home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/torch/csrc/api/include -I/home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/TH -I/home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/THC -I/home/wolf/anaconda3/include/python3.7m -c cpu/graclus.cpp -o build/temp.linux-x86_64-3.7/cpu/graclus.o -DTORCH_API_INCLUDE_EXTENSION_H -DTORCH_EXTENSION_NAME=graclus_cpu -D_GLIBCXX_USE_CXX11_ABI=0 -std=c++11
  cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
  In file included from /home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/ATen/ATen.h:9:0,
                   from /home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/torch/csrc/api/include/torch/types.h:3,
                   from /home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/torch/csrc/api/include/torch/data/dataloader_options.h:4,
                   from /home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/torch/csrc/api/include/torch/data/dataloader.h:3,
                   from /home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/torch/csrc/api/include/torch/data.h:3,
                   from /home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/torch/csrc/api/include/torch/all.h:4,
                   from /home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/torch/extension.h:4,
                   from cpu/graclus.cpp:1:
  cpu/graclus.cpp: In lambda function:
  cpu/graclus.cpp:52:43: error: invalid initialization of reference of type ‘const at::Type&’ from expression of type ‘c10::ScalarType’
     AT_DISPATCH_ALL_TYPES(weight.scalar_type(), "weighted_graclus", [&] {
                           ~~~~~~~~~~~~~~~~~~^~
  /home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/ATen/Dispatch.h:70:32: note: in definition of macro ‘AT_DISPATCH_ALL_TYPES’
       const at::Type& the_type = TYPE;                                         \
                                  ^~~~
  error: command 'gcc' failed with exit status 1
  ----------------------------------------
  ERROR: Failed building wheel for torch-cluster
  Running setup.py clean for torch-cluster
Failed to build torch-cluster
Installing collected packages: torch-cluster
  Running setup.py install for torch-cluster ... error
    ERROR: Complete output from command /home/wolf/anaconda3/bin/python -u -c 'import setuptools, tokenize;__file__='"'"'/tmp/pip-install-19kdi0h_/torch-cluster/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-qz6dlbm0/install-record.txt --single-version-externally-managed --compile:
    ERROR: running install
    running build
    running build_py
    creating build
    creating build/lib.linux-x86_64-3.7
    creating build/lib.linux-x86_64-3.7/test
    copying test/__init__.py -> build/lib.linux-x86_64-3.7/test
    copying test/test_radius.py -> build/lib.linux-x86_64-3.7/test
    copying test/utils.py -> build/lib.linux-x86_64-3.7/test
    copying test/test_sampler.py -> build/lib.linux-x86_64-3.7/test
    copying test/test_rw.py -> build/lib.linux-x86_64-3.7/test
    copying test/test_fps.py -> build/lib.linux-x86_64-3.7/test
    copying test/test_knn.py -> build/lib.linux-x86_64-3.7/test
    copying test/test_grid.py -> build/lib.linux-x86_64-3.7/test
    copying test/test_nearest.py -> build/lib.linux-x86_64-3.7/test
    copying test/test_graclus.py -> build/lib.linux-x86_64-3.7/test
    creating build/lib.linux-x86_64-3.7/torch_cluster
    copying torch_cluster/__init__.py -> build/lib.linux-x86_64-3.7/torch_cluster
    copying torch_cluster/grid.py -> build/lib.linux-x86_64-3.7/torch_cluster
    copying torch_cluster/fps.py -> build/lib.linux-x86_64-3.7/torch_cluster
    copying torch_cluster/nearest.py -> build/lib.linux-x86_64-3.7/torch_cluster
    copying torch_cluster/knn.py -> build/lib.linux-x86_64-3.7/torch_cluster
    copying torch_cluster/rw.py -> build/lib.linux-x86_64-3.7/torch_cluster
    copying torch_cluster/sampler.py -> build/lib.linux-x86_64-3.7/torch_cluster
    copying torch_cluster/graclus.py -> build/lib.linux-x86_64-3.7/torch_cluster
    copying torch_cluster/radius.py -> build/lib.linux-x86_64-3.7/torch_cluster
    running build_ext
    building 'torch_cluster.graclus_cpu' extension
    creating build/temp.linux-x86_64-3.7
    creating build/temp.linux-x86_64-3.7/cpu
    gcc -pthread -B /home/wolf/anaconda3/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include -I/home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/torch/csrc/api/include -I/home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/TH -I/home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/THC -I/home/wolf/anaconda3/include/python3.7m -c cpu/graclus.cpp -o build/temp.linux-x86_64-3.7/cpu/graclus.o -DTORCH_API_INCLUDE_EXTENSION_H -DTORCH_EXTENSION_NAME=graclus_cpu -D_GLIBCXX_USE_CXX11_ABI=0 -std=c++11
    cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
    In file included from /home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/ATen/ATen.h:9:0,
                     from /home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/torch/csrc/api/include/torch/types.h:3,
                     from /home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/torch/csrc/api/include/torch/data/dataloader_options.h:4,
                     from /home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/torch/csrc/api/include/torch/data/dataloader.h:3,
                     from /home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/torch/csrc/api/include/torch/data.h:3,
                     from /home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/torch/csrc/api/include/torch/all.h:4,
                     from /home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/torch/extension.h:4,
                     from cpu/graclus.cpp:1:
    cpu/graclus.cpp: In lambda function:
    cpu/graclus.cpp:52:43: error: invalid initialization of reference of type ‘const at::Type&’ from expression of type ‘c10::ScalarType’
       AT_DISPATCH_ALL_TYPES(weight.scalar_type(), "weighted_graclus", [&] {
                             ~~~~~~~~~~~~~~~~~~^~
    /home/wolf/anaconda3/lib/python3.7/site-packages/torch/lib/include/ATen/Dispatch.h:70:32: note: in definition of macro ‘AT_DISPATCH_ALL_TYPES’
         const at::Type& the_type = TYPE;                                         \
                                    ^~~~
    error: command 'gcc' failed with exit status 1
    ----------------------------------------
ERROR: Command "/home/wolf/anaconda3/bin/python -u -c 'import setuptools, tokenize;__file__='"'"'/tmp/pip-install-19kdi0h_/torch-cluster/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-qz6dlbm0/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-install-19kdi0h_/torch-cluster/

**Comment by rusty1s** [comment]
Are you sure you are running PyTorch 1.2 and not PyTorch 1.0. This looks like a classic PyTorch 1.0 error. If this is not the case, you may try to downgrade your gcc to version 5.4.

**Comment by choltz95** [comment]
Not sure if this should be its own issue, but I had to downgrade torch to 1.1 torchvision to 0.3.0:

```
  File "/prj/mlsoc2/utilization/conda/envs/hmu_v1/lib/python3.6/site-packages/torch/nn/modules/module.py", line 547, in __call__
    result = self.forward(*input, **kwargs)
  File "/prj/mlsoc2/utilization/conda/envs/hmu_v1/lib/python3.6/site-packages/torch_geometric/nn/conv/gcn_conv.py", line 100, in forward
    self.improved, x.dtype)
  File "/prj/mlsoc2/utilization/conda/envs/hmu_v1/lib/python3.6/site-packages/torch_geometric/nn/conv/gcn_conv.py", line 78, in norm
    edge_index, edge_weight, fill_value, num_nodes)
  File "/prj/mlsoc2/utilization/conda/envs/hmu_v1/lib/python3.6/site-packages/torch_geometric/utils/loop.py", line 107, in add_remaining_self_loops
    inv_mask = 1 - mask
  File "/prj/mlsoc2/utilization/conda/envs/hmu_v1/lib/python3.6/site-packages/torch/tensor.py", line 325, in __rsub__
    return _C._VariableFunctions.rsub(self, other)
RuntimeError: Subtraction, the `-` operator, with a bool tensor is not supported. If you are trying to invert a mask, use the `~` or `bitwise_not()` operator instead.
```

**Comment by rusty1s** [comment]
Yes, this is a PyTorch 1.2 issue, but it already fixed in master. Installing PyTorch from
source fixes this (just clone and run `python setup.py install`). I will craft a new minor release soon.

**Comment by messyQ1128** [comment]
> @messyQ1128 make sure you are on cuda 8.0 or 9.2. cuda 9.0 breaks torch with an error that looks similar to the one you posted.
> Note, **if you downgrade torch, you will have to reinstall the torch-geom dependencies.**


Thanks a lot for your  instruction, and the problem has been fixed when i upgrade cuda 9.0 to 10.0.
Thanks a lot!
